vocode/streaming/agent/anthropic_agent.py

```
@@ -61,18 +65,36 @@ async def generate_response(
             self.transcript
         )
 
-        stream = await self.anthropic_async_client.messages.create(
-            max_tokens=1024,
-            messages=messages_formatted,
-            system=self.agent_config.prompt_preamble,
-            model="claude-3-haiku-20240307",
-            stream=True,
-        )
+        delay = initial_delay
+        for i in range(retry_attempts):
+            try:
+                stream = await self.anthropic_async_client.messages.create(
+                    max_tokens=1024,
+                    messages=messages_formatted,
+                    system=self.agent_config.prompt_preamble,
+                    model="claude-3-haiku-20240307",
+                    stream=True,
+                )
 
-        async for message in collate_response_async(
-            anthropic_get_tokens(stream), get_functions=True
-        ):
-            yield BaseMessage(text=message)
+                async for message in collate_response_async(
+                    anthropic_get_tokens(stream), get_functions=True
+                ):
+                    yield BaseMessage(text=message)
+                break
+            except anthropic.APIStatusError as e:
+                message_json = json.loads(
```

This looks a bit brittle. Doesn't the API just return an error code like 529, or you can search the error message for `overloaded_error`? See https://docs.anthropic.com/en/api/errors .

---

not really because it's giving me a `200`. I assume this is a bug on their side?

The message I get in the error is:
`"{'type': 'error', 'error': {'details': None, 'type': 'overloaded_error', 'message': 'Overloaded'}}"`

---

According to [this info](https://github.com/anthropics/anthropic-sdk-python?tab=readme-ov-file#handling-errors) it should return the correct error code, e.g. it should only raise an APIStatusError if e.status_code is not 200. I guess it could be a bug on their side still, but in that case, I would simplify your check to just search the error message string for `overloaded_error`. Also, as I look at the error codes, 429 is the normal one raised by APIs when you are using too much (overloaded sounds more like the actual Antrophic service was loaded, were 429 sounds more like we are using too much of our quota).

It can be helpful to include a comment and link to Antrophic that defines how errors are supposed to be returned, so we can check in the future if the code and their API still syncs.